File: 2011_2012_sujet_3_convergences/TPconv_11.m
%Illustration convergence presque sure
%% Question 1
m=mean(1:6)
sigma=sqrt(mean([1:6].^2) - m^2)
%% Question 2
X=runifd(5000,6);
S=cumsum(X);
S2=S./[1:5000]';
%% Question 3
plot(S2)
hold on;
plot(m*ones(1,5000))
%% Question 4
clf
plot((([1:5000]').^0.1).*(S2-m))
plot([1:5000]'.*(S2-m))
plot(sqrt([1:5000]').*(S2-m))
%% Question 5
clf
plotrep(X(1:100))
hold on
plotrep(X(1:500))
plotrep(X(1:1000))
plotrep(X(1:5000))
stairs([1:6],[1:6]/6,'r')
hold off
%% Question 6-7-8-9
clf
U=rand(1,100);
E=(U<1/2);
V=cumsum(E./2.^[1:100]);
plot(V);
hold on
U=rand(1,100);
E=(U<1/2);
V=cumsum(E./2.^[1:100]);
plot(V,'r');
%% Question 10
clf
for i=1:2000
U=rand(1,100);
E=(U<1/2);
W(i)=sum(E./2.^[1:100]);
end
plotrep(W)
%% Question 12
clf
U=rand(1,100);
E=(U<1/2)+(-1)*(U>1/2);
V=cumsum(E./2.^[1:100]);
plot(V);
hold on
U=rand(1,100);
E=(U<1/2)+(-1)*(U>1/2);
V=cumsum(E./2.^[1:100]);
plot(V,'r');
clf
for i=1:2000
U=rand(1,100);
E=(U<1/2)+(-1)*(U>1/2);
W(i)=sum(E./2.^[1:100]);
end
plotrep(W)
clf
% Theoreme central limite
%% Question 13
clf
stem([200:400],dbinom([200:400],1000,0.3));
hold on;
plot([200:400],dnorm([200:400],1000*0.3,sqrt(1000*0.3*0.7)),'r');
hold off;
%% Question 14 
TCLde(10,1000)
TCLde(50,1000)
TCLde(300,1000)
TCLde(1000,1000)
%% Question 15
clf
B=rbinom(5000,10,2/10);
hold on
plotrep(B)
a=[0:8];
poi=exp(-2)*2.^a./factorial(a);
stairs(a,cumsum(poi),'r')
B=rbinom(5000,100,2/100);
plotrep(B)
B=rbinom(5000,1000,2/1000);
plotrep(B)
%%

File: 2011_2012_sujet_3_convergences/TPconv_11.py
"""Illustration de la convergence — 2011-2012."""
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm, binom as binom_dist

from plotrep import plotrep
from runifd import runifd
from TCLde import TCLde
from Markov import Markov

# %% Question 1
m = np.mean(np.arange(1, 7))
sigma = np.sqrt(np.mean(np.arange(1, 7)**2) - m**2)
print(f'm={m}, sigma={sigma}')

# %% Question 2
X = runifd(5000, 6)
S = np.cumsum(X)
S2 = S / np.arange(1, 5001)

# %% Question 3
plt.figure()
plt.plot(S2); plt.plot(m * np.ones(5000))
plt.title('LGN'); plt.show()

# %% Question 4
plt.figure(); plt.plot((np.arange(1, 5001)**0.1) * (S2 - m)); plt.show()
plt.figure(); plt.plot(np.arange(1, 5001) * (S2 - m)); plt.show()
plt.figure(); plt.plot(np.sqrt(np.arange(1, 5001)) * (S2 - m)); plt.show()

# %% Question 5
plt.figure()
for n in [100, 500, 1000, 5000]:
    plotrep(X[:n])
plt.step(np.arange(1, 7), np.arange(1, 7)/6, 'r', where='post')
plt.show()

# %% Question 6-7-8-9
plt.figure()
U = np.random.rand(100); E = (U < 0.5).astype(float)
V = np.cumsum(E / 2**np.arange(1, 101)); plt.plot(V)
U = np.random.rand(100); E = (U < 0.5).astype(float)
V = np.cumsum(E / 2**np.arange(1, 101)); plt.plot(V, 'r')
plt.show()

# %% Question 10
W = np.zeros(2000)
for i in range(2000):
    U = np.random.rand(100); E = (U < 0.5).astype(float)
    W[i] = np.sum(E / 2**np.arange(1, 101))
plt.figure(); plotrep(W); plt.show()

# %% Question 12
plt.figure()
U = np.random.rand(100); E = (U < 0.5).astype(float) + (-1)*(U > 0.5).astype(float)
V = np.cumsum(E / 2**np.arange(1, 101)); plt.plot(V)
U = np.random.rand(100); E = (U < 0.5).astype(float) + (-1)*(U > 0.5).astype(float)
V = np.cumsum(E / 2**np.arange(1, 101)); plt.plot(V, 'r')
plt.show()

W = np.zeros(2000)
for i in range(2000):
    U = np.random.rand(100); E = (U < 0.5).astype(float) + (-1)*(U > 0.5).astype(float)
    W[i] = np.sum(E / 2**np.arange(1, 101))
plt.figure(); plotrep(W); plt.show()

# %% Question 13 — Moivre-Laplace approximation
k = np.arange(200, 401)
plt.figure()
plt.stem(k, binom_dist.pmf(k, 1000, 0.3), basefmt=' ', markerfmt='b.', linefmt='b-')
x = np.linspace(200, 400, 200)
plt.plot(x, norm.pdf(x, 1000*0.3, np.sqrt(1000*0.3*0.7)), 'r')
plt.show()

# %% Question 14 — TCL
for n_val in [10, 50, 300, 1000]:
    plt.figure(); TCLde(n_val, 1000); plt.title(f'TCL n={n_val}'); plt.show()

# %% Question 15 — Binomial → Poisson
a_arr = np.arange(9)
poi = np.exp(-2) * 2**a_arr / np.array([np.math.factorial(ai) for ai in a_arr])
plt.figure()
for n_val, p_val in [(10, 2/10), (100, 2/100), (1000, 2/1000)]:
    B = np.random.binomial(n_val, p_val, 5000)
    plotrep(B)
plt.step(a_arr, np.cumsum(poi), 'r', where='post')
plt.show()
